File: backend/app/domain.py
"""Core domain types used across the engine, extractors, and API layers."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Literal

import numpy as np


@dataclass(frozen=True)
class OCRWord:
    """A single word returned by the browser's OCR engine."""

    text: str
    confidence: float
    bbox: tuple[float, float, float, float]


@dataclass(frozen=True)
class ImageMeta:
    """Metadata about the source image."""

    width: int
    height: int
    dpi: float | None = None
    orientation: int = 1


@dataclass
class ExtractedField:
    """A single field extracted from the OCR payload by one extractor."""

    name: str
    value: str | None
    bbox: tuple[float, float, float, float] | None
    confidence: float
    evidence_spans: list[tuple[float, float, float, float]] = field(default_factory=list)


Mode = Literal["retail_image", "ecommerce_listing"]
Category = Literal["food", "non_food", "cosmetics", "seeds", "unknown"]


@dataclass(frozen=True)
class ScanContext:
    mode: Mode = "retail_image"
    category: Category = "unknown"
    imported: bool | None = None
    inspection_date: date | None = None


MeasurementMethod = Literal[
    "direct_metadata",
    "geometry_estimate",
    "relative_readability",
    "not_measurable",
]
QualityStatus = Literal[
    "acceptable",
    "usable_with_warnings",
    "retake_recommended",
    "unreadable",
]
VerdictStatus = Literal["pass", "fail", "warn", "manual_review", "na"]
OverallStatus = Literal["pass", "fail", "mixed", "manual_review"]
Severity = Literal["critical", "warning", "info"]


@dataclass
class Verdict:
    rule_id: str
    status: VerdictStatus
    severity: Severity
    citation: str
    evidence: str
    evidence_bboxes: list[tuple[float, float, float, float]]
    failure_message: str | None
    rule_version: str
    confidence: float = 1.0
    reasoning: str = "Legacy verdict"
    measurement_method: MeasurementMethod = "not_measurable"


@dataclass(frozen=True)
class OCRLine:
    word_indexes: tuple[int, ...]
    bbox: tuple[float, float, float, float]
    median_character_height: float


@dataclass(frozen=True)
class VisualMetric:
    name: str
    value: float
    unit: str
    confidence: float
    method: str
    evidence_bboxes: tuple[tuple[float, float, float, float], ...] = ()


@dataclass(frozen=True)
class QualitySummary:
    status: QualityStatus
    score: float
    metrics: tuple[VisualMetric, ...]
    guidance: tuple[str, ...]


@dataclass(frozen=True)
class DecodedImage:
    image: np.ndarray
    width: int
    height: int
    metadata: dict[str, object]


@dataclass(frozen=True)
class PanelEstimate:
    bbox: tuple[float, float, float, float]
    confidence: float
    corners: tuple[tuple[float, float], ...] = ()
    physical_width_mm: float | None = None
    scale_method: MeasurementMethod = "not_measurable"


@dataclass(frozen=True)
class ReadabilityAssessment:
    score: float
    character_height_px: float
    estimated_mm: float | None
    error_mm: float | None
    method: MeasurementMethod
    scale_confidence: float
    status: VerdictStatus
    reasoning: str
    evidence_bboxes: tuple[tuple[float, float, float, float], ...] = ()


@dataclass(frozen=True)
class PlacementResult:
    status: VerdictStatus
    relationship: str
    confidence: float
    reasoning: str
    evidence_bboxes: tuple[tuple[float, float, float, float], ...] = ()


@dataclass(frozen=True)
class AnalysisInput:
    extracted: dict[str, ExtractedField | None]
    quality: QualitySummary
    readability: dict[str, ReadabilityAssessment]
    placement: dict[str, PlacementResult]


@dataclass(frozen=True)
class AnalysisResult:
    quality: QualitySummary
    extracted: dict[str, ExtractedField | None]
    verdicts: tuple[Verdict, ...]
    overall_status: OverallStatus
    analysis_version: str


@dataclass(frozen=True)
class FontSizeBracket:
    max_value: float | None
    normal_mm: float
    blown_mm: float


@dataclass(frozen=True)
class FontSizeTable:
    brackets: list[FontSizeBracket]


@dataclass(frozen=True)
class FontSizeRuleSet:
    key: str
    citation: str
    effective_from: str
    superseded_date: str | None
    table_I: FontSizeTable | None = None
    table_II: FontSizeTable | None = None
    letter_min_mm: float | None = None
    letter_blown_min_mm: float | None = None


@dataclass(frozen=True)
class FontSizeRules:
    versions: dict[str, FontSizeRuleSet]
    default_version: str
    exemption_applies_when_another_law_governs: bool
    exempted_declarations: list[str]
    exempted_categories: list[str]
    enforcement_scale_confidence: float
    boundary_error_policy: Literal["manual_review"]


@dataclass(frozen=True)
class CheckConfig:
    rule_id: str
    citation: str
    field: str
    check_type: str
    severity: Severity
    requires: list[str]
    failure_message: str
    tax_inclusive_phrase_regex: str | None = None
    requires_unit_in: list[str] | None = None
    pin_code_regex: str | None = None
    email_regex: str | None = None
    phone_regex: str | None = None
    date_format_regex: str | None = None
    skipped_when_category_in: list[str] | None = None
    skipped_when_mode: str | None = None


@dataclass(frozen=True)
class ConfidenceThresholds:
    pass_min: float
    warn_min: float


@dataclass(frozen=True)
class RulesConfig:
    version: str
    schema_version: int
    font_size: FontSizeRules
    confidence_thresholds: ConfidenceThresholds
    checks: list[CheckConfig]

    def check_by_id(self, rule_id: str) -> CheckConfig | None:
        for c in self.checks:
            if c.rule_id == rule_id:
                return c
        return None

File: frontend/lib/rules/domain.ts
export type BoundingBox = [number, number, number, number];

export interface OCRWord {
  text: string;
  confidence: number;
  bbox: BoundingBox;
}

export interface ImageMeta {
  width: number;
  height: number;
  dpi?: number | null;
  orientation?: number;
}

export interface ExtractedField {
  name: string;
  value: string | null;
  bbox: BoundingBox | null;
  confidence: number;
  evidence_spans: BoundingBox[];
  conflicting_values?: string[];
}

export type Mode = 'retail_image' | 'ecommerce_listing';
export type Category = 'food' | 'non_food' | 'cosmetics' | 'seeds' | 'unknown';

export interface ScanContext {
  mode?: Mode;
  category?: Category;
  imported?: boolean | null;
  inspection_date?: string | null;
}

export type MeasurementMethod =
  | 'direct_metadata'
  | 'geometry_estimate'
  | 'relative_readability'
  | 'not_measurable';

export type QualityStatus =
  | 'acceptable'
  | 'usable_with_warnings'
  | 'retake_recommended'
  | 'unreadable';

export type VerdictStatus = 'pass' | 'fail' | 'warn' | 'manual_review' | 'na';
export type OverallStatus = 'pass' | 'fail' | 'mixed' | 'manual_review';
export type Severity = 'critical' | 'warning' | 'info';

export interface Verdict {
  rule_id: string;
  status: VerdictStatus;
  severity: Severity;
  citation: string;
  evidence: string;
  evidence_bboxes: BoundingBox[];
  failure_message: string | null;
  rule_version: string;
  confidence: number;
  reasoning: string;
  measurement_method: MeasurementMethod;
}

export interface VisualMetric {
  name: string;
  value: number;
  unit: string;
  confidence: number;
  method: string;
  evidence_bboxes?: BoundingBox[];
}

export interface QualitySummary {
  status: QualityStatus;
  score: number;
  metrics: VisualMetric[];
  guidance: string[];
}

export interface ReadabilityAssessment {
  score: number;
  character_height_px: number;
  estimated_mm: number | null;
  error_mm: number | null;
  method: MeasurementMethod;
  scale_confidence: number;
  status: VerdictStatus;
  reasoning: string;
  evidence_bboxes?: BoundingBox[];
}

export interface PlacementResult {
  status: VerdictStatus;
  relationship: string;
  confidence: number;
  reasoning: string;
  evidence_bboxes?: BoundingBox[];
}

export interface AnalysisInput {
  extracted: Record<string, ExtractedField | null | undefined>;
  quality: QualitySummary;
  readability?: Record<string, ReadabilityAssessment>;
  placement?: Record<string, PlacementResult>;
}

export interface ScanResponse {
  scan_id: number;
  processing_status: 'processing' | 'complete' | 'failed';
  quality: QualitySummary;
  extracted_fields: Record<string, ExtractedField | null>;
  verdicts: Verdict[];
  overall_status: OverallStatus;
  analysis_version: string;
}

export interface ApplicabilityDecision {
  has_required_context: boolean;
  skipped: boolean;
  reasoning: string;
}

export interface FontSizeBracket {
  max_value: number | null;
  normal_mm: number;
  blown_mm: number;
}

export interface FontSizeTable {
  brackets: FontSizeBracket[];
}

export interface FontSizeRuleSet {
  key: string;
  citation: string;
  effective_from: string;
  superseded_date: string | null;
  table_I?: FontSizeTable | null;
  table_II?: FontSizeTable | null;
  letter_min_mm?: number | null;
  letter_blown_min_mm?: number | null;
}

export interface FontSizeRules {
  versions: Record<string, FontSizeRuleSet>;
  default_version: string;
  exemption_applies_when_another_law_governs: boolean;
  exempted_declarations: string[];
  exempted_categories: string[];
  enforcement_scale_confidence: number;
  boundary_error_policy: 'manual_review';
}

export interface AppliesWhenConfig {
  mode_in?: string[];
  category_in?: string[];
  imported?: boolean;
  context_required?: string[];
}

export interface CheckConfig {
  rule_id: string;
  citation: string;
  field: string;
  check_type: string;
  severity: Severity;
  requires: string[];
  failure_message: string;
  tax_inclusive_phrase_regex?: string | null;
  requires_unit_in?: string[] | null;
  pin_code_regex?: string | null;
  email_regex?: string | null;
  phone_regex?: string | null;
  date_format_regex?: string | null;
  skipped_when_category_in?: string[] | null;
  skipped_when_mode?: string | null;
  applies_when?: AppliesWhenConfig;
  placement?: Record<string, unknown>;
  readability?: Record<string, unknown>;
  effective_from?: string | null;
  exemption?: Record<string, unknown>;
}

export interface ConfidenceThresholds {
  pass_min: number;
  warn_min: number;
}

export interface RulesConfig {
  version: string;
  schema_version: number;
  font_size: FontSizeRules;
  confidence_thresholds: ConfidenceThresholds;
  checks: CheckConfig[];
}
